driver-core/src/test/unit/com/mongodb/internal/connection/TestServerListener.java

```diff
@@ -47,7 +48,9 @@ public void serverClosed(final ServerClosedEvent event) {
     @Override
     public void serverDescriptionChanged(final ServerDescriptionChangedEvent event) {
         notNull("event", event);
-        serverDescriptionChangedEvents.add(event);
+        if (shouldPublishChangeEvent(event.getNewDescription(), event.getPreviousDescription())) {
```

Comment:
I'm not clear why this change is necessary. Wouldn't the publisher of the event handle whether the event is published or not?

Reply:
In the new SDAM test that was added, this method is called from ServerListenerMulticaster. SingleServerCluster#serverDescriptionChanged is also called from the multicaster and that method also checks to see if the event is published.

Reply:
OK, I think this is actually a bigger problem than I thought.  If you roll this change back, `Monitoring a standalone connection - suppress update events for equal server descriptions` will fail, but you have to make that test pass by changing something in the production driver code, not here in a test class.  The place to add this check would be `DefaultServer.DefaultServerStateListener#stateChanged`.  I tried that, and all the tests pass.

But that is going to make some other tests fail, in particular:

    DefaultServerSpecification. invalidate should invoke server listeners
    ServerDiscoveryAndMonitoringTest. shouldPassAllOutcomes[Parse logicalSessionTimeoutMinutes from mongoses]
    ServerDiscoveryAndMonitoringTest. shouldPassAllOutcomes[Parse logicalSessionTimeoutMinutes from replica set]

The underlying problem is that we're using the same mechanism to publish server description change events to the application, and to publish server description changed events to the Cluster implementations.  And the rules are different.  The Cluster implementations have to know about all the server description change events, whereas the application does not.

So somehow we need to break that dependency.  I think we need two ServerListener's in DefaultServer: one for the application, and one for the Cluster.  And only the former should have the new check for publishing or not.

Reply:
I have added an application listener in DefaultServer. All tests now pass.